gateway/src/lib.rs

```diff
@@ -1,10 +1,246 @@
 #![no_std]
 
+use core::ops::Deref;
+
+use multiversx_sc::api::KECCAK256_RESULT_LEN;
+
+use crate::constants::*;
+use crate::events::{ContractCallData};
+
 multiversx_sc::imports!();
 
+mod constants;
+mod events;
+mod proxy;
+
 #[multiversx_sc::contract]
-pub trait Gateway
-{
+pub trait Gateway: proxy::ProxyModule + events::Events {
     #[init]
-    fn init(&self) {}
+    fn init(&self, auth_module: &ManagedAddress, chain_id: &ManagedBuffer) {
+        require!(
+            self.blockchain().is_smart_contract(auth_module),
+            "Invalid auth module"
+        );
+
+        self.auth_module().set_if_empty(auth_module);
+        self.chain_id().set_if_empty(chain_id);
+    }
+
+    #[endpoint(callContract)]
+    fn call_contract(
+        &self,
+        destination_chain: ManagedBuffer,
+        destination_contract_address: ManagedBuffer,
+        payload: ManagedBuffer,
+    ) {
+        let caller = self.blockchain().get_caller();
+
+        self.contract_call_event(
+            caller,
+            destination_chain,
+            destination_contract_address,
+            ContractCallData {
+                hash: self.crypto().keccak256(&payload),
+                payload,
+            },
+        );
+    }
+
+    // Can only be called by the appropriate contract address
+    #[endpoint(validateContractCall)]
+    fn validate_contract_call(
+        &self,
+        command_id: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+        source_chain: &ManagedBuffer,
+        source_address: &ManagedBuffer,
+        payload_hash: &ManagedByteArray<KECCAK256_RESULT_LEN>,
+    ) -> bool {
+        let contract_address = &self.blockchain().get_caller();
+
+        let hash = self.get_is_contract_call_approved_key(
+            command_id,
+            source_chain,
+            source_address,
+            contract_address,
+            payload_hash,
+        );
+
+        let valid = self.contract_call_approved().contains(&hash);
+
+        if valid {
+            self.contract_call_approved().remove(&hash);
+
+            self.contract_call_executed_event(command_id);
+        }
+
+        valid
+    }
+
+    // External Functions
+
+    #[endpoint(execute)]
+    fn execute(&self, input: ExecuteInput<Self::Api>) {
+        let message_hash = self.get_message_hash(&input.data);
+
+        let mut allow_operatorship_transfer: bool =
+            self.auth_validate_proof(&message_hash, &input.proof);
+
+        let execute_data: ExecuteData<Self::Api> =
+            ExecuteData::<Self::Api>::top_decode(input.data).unwrap();
```

Comment:
It's a better practice *before* `unwrap` to make sure that if the `unwrap` fails the user will get a verbose error.
You could `.unwrap_or_else(|| sc_panic!(VERBOSE_ERR_MSG))`